src/interpreter/InterpreterNode.h

```diff
@@ -132,23 +257,518 @@ class InterpreterNode {
         preamble = p;
     }
 
-    /** @brief get list of all children */
-    const std::vector<std::unique_ptr<InterpreterNode>>& getChildren() const {
-        return children;
+protected:
+    std::shared_ptr<InterpreterPreamble> preamble = nullptr;
+};
+
+/**
+ * @class InterpreterViewOperation
+ * @brief Interpreter operation that utilizes the index view from underlying relation should inherit from this
+ *        class.
+ */
+class InterpreterViewOperation {
+public:
+    InterpreterViewOperation(size_t id) : viewId(id) {}
+
+    inline size_t getViewId() const {
+        return viewId;
     }
 
-    /** @brief get relation from handle */
-    InterpreterRelation* getRelation() const {
-        assert(relHandle && "No relation cached\n");
-        return (*relHandle).get();
+protected:
+    size_t viewId;
+};
+
+/**
+ * @class InterpreterBinRelOperation
+ * @brief Interpreter operation that involves with two relations should inherit from this class.
+ *        E.g. Swap, Extend
+ */
+class InterpreterBinRelOperation {
+public:
+    InterpreterBinRelOperation(size_t src, size_t target) : src(src), target(target) {}
+
+    inline size_t getSourceId() const {
+        return src;
+    }
+
+    inline size_t getTargetId() const {
+        return target;
     }
 
 protected:
-    enum InterpreterNodeType type;
-    const RamNode* shadow;
-    std::vector<std::unique_ptr<InterpreterNode>> children;
-    RelationHandle* const relHandle;
-    std::vector<size_t> data;
-    std::shared_ptr<InterpreterPreamble> preamble = nullptr;
+    const size_t src;
+    const size_t target;
+};
+
+/**
+ * @class InterpreterNestedOperation
+ * @brief Encode a nested operation for the interpreter node. E.g. Loop, IndexScan
+ */
+class InterpreterNestedOperation {
+public:
+    InterpreterNestedOperation(std::unique_ptr<InterpreterNode> nested) : nested(std::move(nested)) {}
+
+    inline const InterpreterNode* getNestesOperation() const {
+        return nested.get();
+    };
+
+protected:
+    std::unique_ptr<InterpreterNode> nested;
+};
+
+/**
+ * @class InterpreterCondOperation
+ * @brief Encode a conditional operation for the interpreter node. E.g. Exit, Filter
+ */
+class InterpreterCondOperation {
```

Comment:
You could use the full word here.
```suggestion
class InterpreterConditionalOperation {
```